hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/table/action/rollback/TestMergeOnReadRollbackActionExecutor.java

```diff
@@ -140,6 +158,138 @@ public void testMergeOnReadRollbackActionExecutor(boolean isUsingMarkers) throws
+    assert commitMetadata.getPartitionToWriteStats().containsKey(DEFAULT_SECOND_PARTITION_PATH);
+    List<HoodieWriteStat> hoodieWriteStatOptionList = commitMetadata.getPartitionToWriteStats().get(DEFAULT_FIRST_PARTITION_PATH);
+    // Both update and insert record should enter same existing fileGroup due to small file handling
+    assertEquals(1, hoodieWriteStatOptionList.size());
+    assertEquals(generatedFileID, hoodieWriteStatOptionList.get(0).getFileId());
+    // check insert and update numbers
+    assertEquals(2, hoodieWriteStatOptionList.get(0).getNumInserts());
+    assertEquals(1, hoodieWriteStatOptionList.get(0).getNumUpdateWrites());
+
+    List<HoodieWriteStat> secondHoodieWriteStatOptionList = commitMetadata.getPartitionToWriteStats().get(DEFAULT_SECOND_PARTITION_PATH);
+    // All insert should enter one fileGroup
+    assertEquals(1, secondHoodieWriteStatOptionList.size());
+    String fileIdInPartitionTwo = secondHoodieWriteStatOptionList.get(0).getFileId();
+    assertEquals(2, hoodieWriteStatOptionList.get(0).getNumInserts());
+
+    /**
+     * Rollback
+     */
+    HoodieInstant rollBackInstant = new HoodieInstant(HoodieInstant.State.INFLIGHT, HoodieTimeline.DELTA_COMMIT_ACTION, "002");
+    BaseRollbackPlanActionExecutor mergeOnReadRollbackPlanActionExecutor =
+        new BaseRollbackPlanActionExecutor(context, cfg, table, "003", rollBackInstant, false,
+            cfg.shouldRollbackUsingMarkers());
+    mergeOnReadRollbackPlanActionExecutor.execute().get();
+    MergeOnReadRollbackActionExecutor mergeOnReadRollbackActionExecutor = new MergeOnReadRollbackActionExecutor(
+        context,
+        cfg,
+        table,
+        "003",
+        rollBackInstant,
+        true,
+        false);
+
+    //3. assert the rollback stat
+    Map<String, HoodieRollbackPartitionMetadata> rollbackMetadata = mergeOnReadRollbackActionExecutor.execute().getPartitionMetadata();
+    assertEquals(2, rollbackMetadata.size());
+
+    //4. assert filegroup after rollback, and compare to the rollbackstat
+    // assert the first partition data and log file size
+    HoodieRollbackPartitionMetadata partitionMetadata = rollbackMetadata.get(DEFAULT_FIRST_PARTITION_PATH);
+    assertTrue(partitionMetadata.getSuccessDeleteFiles().isEmpty());
+    assertTrue(partitionMetadata.getFailedDeleteFiles().isEmpty());
+    assertEquals(1, partitionMetadata.getRollbackLogFiles().size());
+    assertEquals(1, partitionMetadata.getWrittenLogFiles().size());
+
+    // assert the second partition data and log file size
+    partitionMetadata = rollbackMetadata.get(DEFAULT_SECOND_PARTITION_PATH);
+    assertEquals(1, partitionMetadata.getSuccessDeleteFiles().size());
+    assertTrue(partitionMetadata.getFailedDeleteFiles().isEmpty());
+    assertTrue(partitionMetadata.getRollbackLogFiles().isEmpty());
+    assertEquals(1, partitionMetadata.getSuccessDeleteFiles().size());
```

Comment:
Can you add a read query at the end to validate that hudi returns only all valid records after this rollback.